write-fonts/src/tables/glyf.rs

```diff
@@ -327,12 +372,169 @@ impl FontWrite for SimpleGlyph {
             .for_each(|flag| flag.write_into(writer));
         deltas.iter().for_each(|(_, x, _)| x.write_into(writer));
         deltas.iter().for_each(|(_, _, y)| y.write_into(writer));
+        writer.ensure_word_aligned();
     }
 }
 
 impl crate::validate::Validate for SimpleGlyph {
-    fn validate_impl(&self, _ctx: &mut crate::codegen_prelude::ValidationCtx) {
-        // pass
+    fn validate_impl(&self, ctx: &mut crate::codegen_prelude::ValidationCtx) {
+        if self._instructions.len() > u16::MAX as usize {
+            ctx.report("instructions len overflows");
+        }
+    }
+}
+
+impl Component {
+    /// Create a new component.
+    pub fn new(
+        glyph: GlyphId,
+        anchor: Anchor,
+        transform: Transform,
+        flags: impl Into<ComponentFlags>,
+    ) -> Self {
+        Component {
+            glyph,
+            anchor,
+            flags: flags.into(),
+            transform,
+        }
+    }
+    /// Compute the flags for this glyph, excepting `MORE_COMPONENTS` and
+    /// `WE_HAVE_INSTRUCTIONS`, which must be set manually
+    fn compute_flag(&self) -> CompositeGlyphFlags {
+        self.anchor.compute_flags() | self.transform.compute_flags() | self.flags.into()
+    }
+
+    /// like `FontWrite` but lets us pass in the flags that must be determined
+    /// externally (WE_HAVE_INSTRUCTIONS and MORE_COMPONENTS)
+    fn write_into(&self, writer: &mut crate::TableWriter, extra_flags: CompositeGlyphFlags) {
+        let flags = self.compute_flag() | extra_flags;
+        flags.bits().write_into(writer);
+        self.glyph.write_into(writer);
+        self.anchor.write_into(writer);
+        self.transform.write_into(writer);
+    }
+}
+
+impl CompositeGlyph {
+    /// Create a new composite glyph, with the provided component.
+    ///
+    /// Additional components can be added with [`add_component`][Self::add_component]
```

Comment:
why special case one component per composite glyph? I'd have expected CompositeGlyph::new to take one or more components (a slice of len > 1 probably)

Reply:
It's possible I'm being overly clever here, but the basic idea is to enforce the invariant that a composite glyph always has at least one component. There's no obvious way how else I would express this in rust without getting very fancy. I don't really want to use a slice since I don't know how the caller is storing components (it is likely to be in something like a hashmap, though) and it's also hard to express the idea of having the constructor take a non-empty iterator. I definitely expect to see this API get tweaked as we start to actually try to use it for things, though.

Reply:
`new(vec<Component>...) -> Result<Self, SomeErrorT>` would be much less annoying to work with.

Reply:
Actually ... as with pens this would arguably be nice as an iterator adapter. Collecting into a Composite would be quite nice as long as the error handling is tidy.

Reply:
#251 shows a sketch of suggestion above.